edx/blockstore
# [WIP] Revert "Revert "Upgrade blockstore to use Python 3.8 and Django 2.2+. (#82)""

This PR rolls back the [rollback](https://github.com/edx/blockstore/pull/84) of [Python 3.8 compatibility](https://github.com/edx/blockstore/pull/82), while changing the requirements to allow 3.5 to still run and with the added ability to run both Python versions in the devstack by specifying the variable when invoking make.

**Merge deadline**: Soon's good.

**Testing instructions**:

Blow away your cached images for blockstore and the cached volumes. If you're feeling frisky, do `make dev.up` in devstack to make sure your other devstack containers are in active use, make sure blockstore is stopped, and run `docker system prune` to blow away everything else. If you've got other containers you care about, don't do that. Instead, the following may be enough:

```
docker volume ls |grep blockstore|awk '{print $2}'|xargs docker volume rm
```

This command will probably error out, giving you a list of volumes that can't be destroyed because they're attached to containers. `docker kill`, then `docker rm` each container hash it mentions, then run the above command again.

You may need to manually build both versions of the devstack to make sure old copies aren't present:

```
docker-compose -f docker-compose-3.5.yml build --no-cache
docker-compose -f docker-compose-3.8.yml build --no-cache
```

Now you can launch makeserver or easyserver by specifying an environment variable. For instance:

```
PYTHON_VERSION=3.5 make easyserver
```
...or

```
PYTHON_VERSION=3.8 make testserver
```

And, of course, other relevent commands are also updated:

```
PYTHON_VERSION=3.5 make blockstore-shell
```

From within each shell, you'll want to run

```
make test
```

Unfortunately the containers can't have the same name. Sad. So if you want to run the integration tests with 3.5 (3.8 is the new default), you'll need to use a modified version of the integration test commands:

```
EDXAPP_BLOCKSTORE_API_URL='http://edx.devstack.blockstore.old-test:18251/api/v1/'  EDXAPP_RUN_BLOCKSTORE_TESTS=1 python -Wd -m pytest --ds=cms.envs.test openedx/core/lib/blockstore_api/ openedx/core/djangolib/tests/test_blockstore_cache.py openedx/core/djangoapps/content_libraries/tests/
EDXAPP_BLOCKSTORE_API_URL='http://edx.devstack.blockstore.old-test:18251/api/v1/'  EDXAPP_RUN_BLOCKSTORE_TESTS=1 python -Wd -m pytest --ds=lms.envs.test openedx/core/lib/blockstore_api/ openedx/core/djangolib/tests/test_blockstore_cache.py openedx/core/djangoapps/content_libraries/tests/

```

**Author notes and concerns**:

I don't know if it's worth updating the README for all of this since we don't expect it to last long and we expect most people to be developing on 3.8 in the interim, which the Makefile will use by default. But I can if needed.

**Reviewers**
